### PR TITLE
Refactor devices changed query to pull less from DB

### LINKED ISSUE #5553: `device_lists_stream` table is large and gets queried a lot
`device_lists_stream` is one of the busiest tables in terms of number of rows fetched from it. This turns out to be because clients can ask us to do an unbounded fetch for all devices that have changed between two tokens, which returns a lot if the tokens are *old*.

We also store a row per device, but we don't actually seem to read the device.

### PR DESCRIPTION
**Pull request by erikjohnston**
When a client asks for users whose devices have changed since a token we
used to pull *all* users from the database since the token, which could
easily be thousands of rows for old tokens.

This PR changes this to only check for changes for users the client is
actually interested in.

Fixes #5553

### CODE REVIEW

File: synapse/handlers/sync.py [hunk 1]
```diff
@@ -1076,21 +1072,23 @@ def _generate_sync_entry_for_device_list(
...
+                user_id
+            )
+
             # TODO: Check that these users are actually new, i.e. either they
             # weren't in the previous sync *or* they left and rejoined.
-            changed.update(newly_joined_or_invited_users)
-
-            if not changed and not newly_left_users:
-                defer.returnValue(DeviceLists(changed=[], left=newly_left_users))
+            changed = users_who_share_room & set(newly_joined_or_invited_users)
```
Comment (richvdh) [inline synapse/handlers/sync.py]: this used to be a union: is the change to & deliberate?

File: synapse/handlers/sync.py [hunk 2]
```diff
@@ -1076,21 +1072,23 @@ def _generate_sync_entry_for_device_list(
...
-                user_id
+            changed_users = yield self.store.get_user_whose_devices_changed(
+                since_token.device_list_key, users_who_share_room
             )
 
+            changed.update(changed_users)
+
             defer.returnValue(
                 DeviceLists(
-                    changed=users_who_share_room & changed,
```
Comment (richvdh) [inline synapse/handlers/sync.py]: argh no, I really don't understand how this used to work or how it works now. wtf is `changed`? why does it have a shit name? why did we previously intersect with `users_who_share_room` and now don't? and similar questions.

File: synapse/storage/devices.py [hunk 3]
```diff
@@ -391,22 +391,53 @@ def _get_devices_with_keys_by_user_txn(self, txn, user_id):
...
 
-    @defer.inlineCallbacks
-    def get_user_whose_devices_changed(self, from_key):
-        """Get set of users whose devices have changed since `from_key`.
+    def get_user_whose_devices_changed(self, from_key, user_ids):
+        """Get set of users whose devices have changed since `from_key` that
+        are in the given list of user_ids.
+
+        Args:
+            user_ids (Iterable[str])
```
Comment (richvdh) [inline synapse/storage/devices.py]: these are in the wrong order...

File: synapse/storage/devices.py [hunk 4]
```diff
@@ -391,22 +391,53 @@ def _get_devices_with_keys_by_user_txn(self, txn, user_id):
...
+            from_key: The device lists stream token
```
Comment (richvdh) [inline synapse/storage/devices.py]:
```suggestion
            from_key (str): The device lists stream token
```

File: synapse/storage/devices.py [hunk 5]
```diff
@@ -391,22 +391,53 @@ def _get_devices_with_keys_by_user_txn(self, txn, user_id):
...
+        to_check = list(
+            self._device_list_stream_cache.get_entities_changed(user_ids, from_key)
+        )
+
+        if not to_check:
+            return defer.succeed(set())
+
+        # We now check the database for all users in `to_check`, in batches.
+        batch_size = 100
+        chunks = [
```
Comment (richvdh) [inline synapse/storage/devices.py]: `batch_iter` is a thing.

File: synapse/handlers/device.py [hunk 6]
```diff
@@ -188,10 +192,6 @@ def get_user_ids_changed(self, user_id, from_token):
                         break
 
         if possibly_changed or possibly_left:
-            users_who_share_room = yield self.store.get_users_who_share_room_with_user(
-                user_id
-            )
-
             # Take the intersection of the users whose devices may have changed
             # and those that actually still share a room with the user
             possibly_joined = possibly_changed & users_who_share_room
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/handlers/device.py]: Possibly. We do go and add entries in the loop, though they should all be new users

**Comment by erikjohnston** [comment]
@richvdh I've shuffled the code a bit and added comments to try and explain what's going on in the `/sync` code. I'm fairly sure they were functionally equivalent to before, but there was a lot of moving the intersection up before function calls and the like. Hopefully it should maybe make a bit of sense now.

File: synapse/storage/devices.py [hunk 7]
```diff
@@ -391,22 +392,49 @@ def _get_devices_with_keys_by_user_txn(self, txn, user_id):
...
+
+            sql = """
+                SELECT DISTINCT user_id FROM device_lists_stream
+                WHERE stream_id > ?
+                AND user_id IN (%s)
+            """
+
+            for chunk in batch_iter(to_check, 100):
+                txn.execute(
+                    sql % (",".join("?" for _ in chunk),), [from_key] + list(chunk)
```
Comment (richvdh) [inline synapse/storage/devices.py]: `chunk` is a tuple, so copying to an intermediate list is redundant:

```suggestion
                    sql % (",".join("?" for _ in chunk),), (from_key,) + chunk
```

File: synapse/handlers/sync.py [hunk 8]
```diff
@@ -1058,40 +1058,74 @@ def _generate_sync_entry_for_device_list(
...
+            )
+
+            # Step 1, check for changes in devices of users we share a room with
+            users_that_have_changed = yield self.store.get_users_whose_devices_changed(
+                since_token.device_list_key, users_who_share_room
             )
 
-            # TODO: Be more clever than this, i.e. remove users who we already
-            # share a room with?
+            # Step 2, check for newly joined rooms
```
Comment (richvdh) [inline synapse/handlers/sync.py]: errm, I think this is a different step 2 to that mentioned at line 1097. Perhaps you want `1a` and `1b`.

**Pull request merged**